### ISSUE TITLE
Более развёрнутая информация по Liquibase

### ISSUE THREAD

**Issue by grfoger** [peterarsentev/job4j_features_bugs#1128]
Добрый день.
Хотелось бы в курсе чуть больше информации о работе и настройке Liquibase.
-Хотелось бы знать что он может и как он должен работать. (В курсе он используется для тестов, а для загрузки в БД используется через вызов плагина или сбор билда. Мне например казалось что он должен обновлять информацию при запуске приложения)
-Хотелось бы знать способы настройки (pom, properties, yaml), как и на что они влияют и в каких всё это отношениях со Spring Boot. Какие практики хороши, какие использовать не стоит и почему.
-Хотелось бы иметь базовые примеры работающих настроек под связки  Hibernate + Liqubase + H2 + PostgreSQL  с возможностью писать скрипты в формате sql.
-Глубокая информация о миграциях БД скорее всего кмк была бы излишней, но было бы неплохо получить о ней обзорное представление. На данный момент это скорее тёмная лошадка из разряда "скопировал - вставил - вроде работает".

Всё вышеизложенное ни в коем случае не претензия (курс - шикарный), а просто пожелание  - предложение.

**Comment by grfoger** [comment]
Для меня основной затык был в том что я не понимал что надо изменения для Lb прописывать каждый раз в новом файле и вносить в конфигурацию. Хотелось бы чуть больше акцента на таких пошаговых этапах.

**Comment by ik87** [comment]
```Hibernate + Liqubase + H2 + PostgreSQL```
Вот тут начинается интересное. Это конечно мое скромное имхо. Но раз юзается Hibernate, то вероятно использование Миграции немного бессмысленно, т.к таблицы мы создаем через ОРМ и анатации, Разве что для заполнения какимто методанными...

**Comment by peterarsentev** [comment]
> `Hibernate + Liqubase + H2 + PostgreSQL` Вот тут начинается интересное. Это конечно мое скромное имхо. Но раз юзается Hibernate, то вероятно использование Миграции немного бессмысленно, т.к таблицы мы создаем через ОРМ и анатации, Разве что для заполнения какимто методанными...

А как у вас поддерживается схема?

**Comment by ik87** [comment]
А так, это очень удобно. В моем проекте не используется Hibernate. Тут как бы все очивидно. Есть ДАО в котором описывается SQL запрос, где мы берем данные из таблиц. 
Миграция же решает существенную роль, как бы происходит безшовное обновление БД, разных таблиц, каких то данных (методанных как правило). Т.е не нужно замарачиваться и вручную создавать таблицы, потом их как то вручную внедрять

**Comment by ik87** [comment]
> > `Hibernate + Liqubase + H2 + PostgreSQL` Вот тут начинается интересное. Это конечно мое скромное имхо. Но раз юзается Hibernate, то вероятно использование Миграции немного бессмысленно, т.к таблицы мы создаем через ОРМ и анатации, Разве что для заполнения какимто методанными...
> 
> А как у вас поддерживается схема?

Вот пример:

```
import org.springframework.data.r2dbc.core.DatabaseClient;
import org.springframework.stereotype.Repository;
import reactor.core.publisher.Flux;

@Repository
public class YourRepository {
    private final DatabaseClient databaseClient;

    public YourRepository(DatabaseClient databaseClient) {
        this.databaseClient = databaseClient;
    }

    public Flux<YourEntity> findByCustomQuery(String yourValue) {
        return databaseClient.sql("SELECT * FROM your_table WHERE your_column = :value")
                .bind("value", yourValue)
                .map((row, metadata) -> yourMapperMethod(row)) // Ваш метод для маппинга строки на объект
                .all();
    }
}
```

**Comment by ik87** [comment]
Схемы и т.д ложится уже на миграцию, Причем не допускается удалять changelog т.е если мы хотим что то исправить, делаем следующий коммит, где вносятся правки
Примерный вид

```
//git commit "first commit"
/resource/changelog/schema/2023-01-01-001-init-user-table.sql
/resource/changelog/schema/2023-01-01-002-alter-user-table.sql

//git commit "second commit"
/resource/changelog/schema/2023-01-01-003-delete-user-table.sql
...

```

**Comment by ik87** [comment]
В общем подытожить хотелось бы так. Ликвибейс помогает избавится от головняка, и ни делать ничего вручную в БД, все делается через код и после выкатки в ДЕВ или ПРОД у нас будет то что мы ожидаем

**Comment by ik87** [comment]
Тестирование это отдельная песня. Я выробатал такой момент, для тестов, в рессурсах не создаю отдельно ликвибейс (те в src/test/resources). А юзаю ту которая в main, что бы было 1 в 1, единственное инсертю какие то тестовые данные, но это уже отдельные скрипты.

**Comment by MichaelZalogin** [comment]
> Схемы и т.д ложится уже на миграцию, Причем не допускается удалять changelog т.е если мы хотим что то исправить, делаем следующий коммит, где вносятся правки Примерный вид
> 
> ```
> //git commit "first commit"
> /resource/changelog/schema/2023-01-01-001-init-user-table.sql
> /resource/changelog/schema/2023-01-01-002-alter-user-table.sql
> 
> //git commit "second commit"
> /resource/changelog/schema/2023-01-01-003-delete-user-table.sql
> ...
> ```

У нас есть ченжсеты с параметром run_on_change которые автоматически перезапускаются в случае, если внесены изменения. Часто есть вещи которые постоянно меняются в базе.

**Comment by ik87** [comment]
> > Схемы и т.д ложится уже на миграцию, Причем не допускается удалять changelog т.е если мы хотим что то исправить, делаем следующий коммит, где вносятся правки Примерный вид
> > ```
> > //git commit "first commit"
> > /resource/changelog/schema/2023-01-01-001-init-user-table.sql
> > /resource/changelog/schema/2023-01-01-002-alter-user-table.sql
> > 
> > //git commit "second commit"
> > /resource/changelog/schema/2023-01-01-003-delete-user-table.sql
> > ...
> > ```
> 
> У нас есть ченжсеты с параметром run_on_change которые автоматически перезапускаются в случае, если внесены изменения. Часто есть вещи которые постоянно меняются в базе.

т.е если ты изменил что то в существующем changelog-e то происходит магия с обновлениями ? При откатывании запутаться не получиться ?

**Comment by MichaelZalogin** [comment]
Ченжсеты пишем сразу в xml ченжлоге а не указываем пути к отдельным файлам
     
```
<changeSet id="ххххх" author="хххх" runOnChange="true">
        <comment>some comment</comment>
        <sql>
        </sql>
    </changeSet>
```

**Comment by MichaelZalogin** [comment]
> > > Схемы и т.д ложится уже на миграцию, Причем не допускается удалять changelog т.е если мы хотим что то исправить, делаем следующий коммит, где вносятся правки Примерный вид
> > > ```
> > > //git commit "first commit"
> > > /resource/changelog/schema/2023-01-01-001-init-user-table.sql
> > > /resource/changelog/schema/2023-01-01-002-alter-user-table.sql
> > > 
> > > //git commit "second commit"
> > > /resource/changelog/schema/2023-01-01-003-delete-user-table.sql
> > > ...
> > > ```
> > 
> > 
> > У нас есть ченжсеты с параметром run_on_change которые автоматически перезапускаются в случае, если внесены изменения. Часто есть вещи которые постоянно меняются в базе.
> 
> т.е если ты изменил что то в существующем changelog-e то происходит магия с обновлениями ? При откатывании запутаться не получиться ?

Из за этого чсы не раздуваются в простыню. У нас например в базе надо часто включать или выключать какие то настройки. Важно помнить что перезапускаемые скрипты должны быть UPDATE или DELETE+INSERT

**Comment by ik87** [comment]
мы ченджсеты указываем в changelog-ах через специальные коментарии к sql коду

**Comment by ik87** [comment]
> > > > Схемы и т.д ложится уже на миграцию, Причем не допускается удалять changelog т.е если мы хотим что то исправить, делаем следующий коммит, где вносятся правки Примерный вид
> > > > ```
> > > > //git commit "first commit"
> > > > /resource/changelog/schema/2023-01-01-001-init-user-table.sql
> > > > /resource/changelog/schema/2023-01-01-002-alter-user-table.sql
> > > > 
> > > > //git commit "second commit"
> > > > /resource/changelog/schema/2023-01-01-003-delete-user-table.sql
> > > > ...
> > > > ```
> > > 
> > > 
> > > У нас есть ченжсеты с параметром run_on_change которые автоматически перезапускаются в случае, если внесены изменения. Часто есть вещи которые постоянно меняются в базе.
> > 
> > 
> > т.е если ты изменил что то в существующем changelog-e то происходит магия с обновлениями ? При откатывании запутаться не получиться ?
> 
> Из за этого чсы не раздуваются в простыню. У нас например в базе надо часто включать или выключать какие то настройки. Важно помнить что перезапускаемые скрипты должны быть UPDATE или DELETE+INSERT

отктываться приходилось ? Не происходило непредвиденных проблемм ?

**Comment by MichaelZalogin** [comment]
> > > > > Схемы и т.д ложится уже на миграцию, Причем не допускается удалять changelog т.е если мы хотим что то исправить, делаем следующий коммит, где вносятся правки Примерный вид
> > > > > ```
> > > > > //git commit "first commit"
> > > > > /resource/changelog/schema/2023-01-01-001-init-user-table.sql
> > > > > /resource/changelog/schema/2023-01-01-002-alter-user-table.sql
> > > > > 
> > > > > //git commit "second commit"
> > > > > /resource/changelog/schema/2023-01-01-003-delete-user-table.sql
> > > > > ...
> > > > > ```
> > > > 
> > > > 
> > > > У нас есть ченжсеты с параметром run_on_change которые автоматически перезапускаются в случае, если внесены изменения. Часто есть вещи которые постоянно меняются в базе.
> > > 
> > > 
> > > т.е если ты изменил что то в существующем changelog-e то происходит магия с обновлениями ? При откатывании запутаться не получиться ?
> > 
> > 
> > Из за этого чсы не раздуваются в простыню. У нас например в базе надо часто включать или выключать какие то настройки. Важно помнить что перезапускаемые скрипты должны быть UPDATE или DELETE+INSERT
> 
> отктываться приходилось ? Не происходило непредвиденных проблемм ?

нет, проблем нет. У нас четыре ченжлога. Первый исполняет три других. Первый из этих трех мастер с генеральными схемами, второй настройки в зависимости от выбранного спринг профиля, третий post мастер

**Comment by ik87** [comment]
> нет, проблем нет. У нас четыре ченжлога. Первый исполняет три других. Первый из этих трех мастер с генеральными схемами, второй настройки в зависимости от выбранного спринг профиля, третий post мастер

Видимо сколько контор столько и разных вариантов.
Хибернет юзаете ?

**Comment by MichaelZalogin** [comment]
> > нет, проблем нет. У нас четыре ченжлога. Первый исполняет три других. Первый из этих трех мастер с генеральными схемами, второй настройки в зависимости от выбранного спринг профиля, третий post мастер
> 
> Видимо сколько контор столько и разных вариантов. Хибернет юзаете ?

Нет, у нас проект завязан на аналитику и большой объем данных, много нативного SQL, поэтому у нас JOOQ. Мне больше хибернейта зашел и работает он гораздо быстрее
